src/cmd/watch.rs

```diff
@@ -87,11 +89,15 @@ fn find_gitignore(book_root: &Path) -> Option<PathBuf> {
 }
 
 fn filter_ignored_files(ignore: Gitignore, paths: &[PathBuf]) -> Vec<PathBuf> {
+    let current_dir = env::current_dir().expect("Unable to determine the current directory");
+
     paths
         .iter()
         .filter(|path| {
+            let relative_path =
+                diff_paths(&current_dir, &path).expect("One of the paths should be an absolute");
```

Comment:
IIUC, the fundamental problem is that the original root path might have path components like `.` or `..`, and the paths we're iterating over sometimes don't have those. Since `ignore` is wanting the path prefixes to match exactly, it is failing.

The solution here is to always use relative paths which `ignore` seems to be happier with.

We can't use `strip_prefix` because it doesn't handle `..` very well.

Do I understand the problem correctly? Were there other considerations you made for why you chose this approach?

Can you maybe add a unittest for `filter_ignored_files` that includes a few wonky examples (for example, files outside the gitignore root, gitignore root with `.` or `..` components, where the given paths don't have those, or paths that do have `.` or `..`). For example, on Windows you could have a gitignore root of something like `C:\\example\\src\\../doc/user-guide"` and a changed path that looks like `C:\\example\\src\\../doc/user-guide\\src\\index.md`.  I would imagine these tests have a few different ignore patterns (like `*.tmp` or `src/foo.md`) and check how it behaves with paths outside the ignore root.

Reply:
Sorry for didn't explain it clearly, I think the fundamental problem is [`ignore .matched_path_or_any_parents`](https://github.com/BurntSushi/ripgrep/blob/7099e174acbcbd940f57e4ab4913fee4040c826e/crates/ignore/src/gitignore.rs#L220-L242) input requires a relative path to where(`current dir`) you execute the `mdbook` command.


In this change, what I have tested with `rustup/doc` locally:
```bash
# same path level with book
mdbook serve . --open

# outer path level with book
mdbook serve ./user-guide --open

# inner path level with book
mdbook serve ../doc/user-guide --open

# absolute path
mdbook serve /Users/xxx/Desktop/rustup/doc/user-guide/ --open

# relative path
mdbook serve ../newfolder/../rustup/doc/user-guide  --open
```

Reply:
update in b0a001c